### PR TITLE
Notes convert to Pandas-compat

### LINKED ISSUE #11460: [FEA] Move pandas-related notes from Notes to pandas-compat
**Is your feature request related to a problem? Please describe.**
#8643 introduced a new rST directive `pandas-compat` to our documentation that allows us to collect all documentation relating to differences between cuDF methods and pandas methods. That directive enables us to much more effectively indicate to users when our behavior differs from pandas. However, we have not since made much use of this directive. Many docstrings still contain `Notes` sections that are essentially just descriptions of how a class/function/method differs from pandas.

**Describe the solution you'd like**
We should convert all pandas-related `Notes` sections of docstrings into `pandas-compat` sections. Doing so should be quite straightforward by simply searching for the header `Notes` in all files containing public APIs. For examples on how to use the `pandas-compat` directive, see #8643.

**Describe alternatives you've considered**
None

**Additional context**
There are likely many other places where we differ from pandas that are either 1) undocumented or 2) documented in different ways. We should also aim to improve that documentation. However, that is a nebulous target since it's not immediately clear where all those places are, so this issue is just focusing on the more specific goal of replacing `Notes` sections, which can be easily searched for and replaced as needed.

### PR DESCRIPTION
**Pull request by Touutae-lab**
# Description
Closes #11460
 
image

**I already change and test documentation in my localhost but some of it can't be change due**
> ### BUG
> can't convert Note froms `DataFrame.mode`
> image
>
> convert to 
> image
>
> ```
> Extension error (numpydoc.numpydoc):
> Handler <function mangle_signature at 0x7fe25c23d630> for event 'autodoc-process-signature' threw an exception (exception: 
> Error parsing See Also entry '.. pandas-compat::' in the docstring of mode in /home/touutae/miniconda3/envs/cudf_dev
> /lib/python3.10/contextlib.py.)
> make: *** [Makefile:20: html] Error 2
> ```
>
> it only happened when edited `DataFrame.mode`
>
> _Originally posted by @Touutae-lab in https://github.com/rapidsai/cudf/issues/11460#issuecomment-1407660086_

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by shwina** [comment]
Hi @Touutae-lab -- thanks for this PR!

Regarding the issue you're seeing with `DataFrame.mode`, I can reproduce the same problem. I was able to work around the issue by moving the `.. pandas-compat` section to after the `Examples` section.

Without this workaround, numpydoc/sphinx seems to want to parse the `pandas-compat` section as part of the `See Also` section. Perhaps @isVoid can guess why this is the case, or otherwise it seems like we have a bug in our implementation of the `pandas-compat` extension.

For this PR, can you make it a policy to always include `pandas-compat` as the **final** section when it is used in a docstring?

**Comment by Touutae-lab** [comment]
Hi @shwina,
I just realized when we add `pandas-compat` to some part such as `groupby.GroupBy._reduce` in  can lead to confusion

It will lead to every _VALID_REDUCTIONS documentation.
```
  _VALID_REDUCTIONS = {
        "sum",
        "prod",
        "idxmin",
        "idxmax",
        "min",
        "max",
        "mean",
        "median",
        "nunique",
        "first",
        "last",
        "var",
        "std",
    }
```
for example `idxmin`
image

every page above and that valid in ` _VALID_REDUCTIONS`  rule will contain the same.
image

and result in [`pandas-compat-list` pages](https://docs.rapids.ai/api/cudf/nightly/user_guide/pandascompat) with the same text with many duplication.


**Do you have any suggestion how should i handle this ? or just skipp**

File: python/cudf/cudf/core/groupby/groupby.py [hunk 1]
```diff
@@ -1839,6 +1840,46 @@ def __getitem__(self, key):
             group_keys=self._group_keys,
         )
 
+    def _reduce(
```
Comment (Touutae-lab) [inline python/cudf/cudf/core/groupby/groupby.py:1843]: I did overload methods because i want [This page](https://docs.rapids.ai/api/cudf/nightly/user_guide/pandascompat)
to become more readable.
like this
image

instead of using same ClassName same methods.

**Review by Touutae-lab — COMMENTED** [review]
Overload `_reduce` method in `DataFrameGroupBy`, `SeriesFrameGroupBy` for Generate more relatable `Pandascompat` pages

**Comment by Touutae-lab** [comment]
@shwina I already change what you have mention. can you check again. Thank for your time ! this is my first time contribute in project <3

File: python/cudf/cudf/core/column/lists.py [hunk 2]
```diff
@@ -655,6 +650,11 @@ def sort_values(
         1         [2.0, 8.0, 8.0]
         2              [1.0, 2.0]
         dtype: list
+
+        .. pandas-compat::
+            **list.ListMethods.sort_values**
+
+            * Not supporting: `inplace`, `kind`
```
Comment (wence-) [inline python/cudf/cudf/core/column/lists.py:657]: Let's reword this to
```suggestion
            The ``inplace`` and ``kind`` arguments are currently not supported.
```

File: python/cudf/cudf/core/column/string.py [hunk 3]
```diff
@@ -3537,6 +3533,15 @@ def count(self, pat: str, flags: int = 0) -> SeriesOrIndex:
...
         Int64Index([0, 0, 2, 1], dtype='int64')
+
+        .. pandas-compat::
+            **StringMethods.count**
+
+            -   `flags` parameter currently only supports re.DOTALL
+                and re.MULTILINE.
+            -   Some characters need to be escaped when passing
+                in pat. e.g. ``'$'`` has a special meaning in regex
+                and must be escaped when finding this literal character.
```
Comment (wence-) [inline python/cudf/cudf/core/column/string.py:3544]: This second point is not a pandas compatibility note, but rather just a general note about regex matching, can we therefore restore it to the previous notes section.

File: python/cudf/cudf/core/dataframe.py [hunk 4]
```diff
@@ -5941,7 +5943,7 @@ def mode(self, axis=0, numeric_only=False, dropna=True):
         Returns
         -------
         DataFrame
-            The modes of each column or row.
+            The modes of each column or33 row.
```
Comment (wence-) [inline python/cudf/cudf/core/dataframe.py:5946]:
```suggestion
            The modes of each column or row.
```

File: python/cudf/cudf/core/dataframe.py [hunk 5]
```diff
@@ -6528,9 +6531,10 @@ def to_struct(self, name=None):
         name: optional
             Name of the resulting Series
 
-        Notes
-        -----
-        Note that a copy of the columns is made.
+        .. pandas-compat::
+            **DataFrame.to_struct**
+
+            Note: that a copy of the columns is made.
```
Comment (wence-) [inline python/cudf/cudf/core/dataframe.py:6537]: I think that pandas dataframes don't support a `to_struct` method (they don't have struct dtypes because its just an object), so this is (I think) not a pandas compatibility note.

File: python/cudf/cudf/core/dataframe.py [hunk 6]
```diff
@@ -6705,6 +6698,18 @@ def append(
...
+            **DataFrame.append**
+
+            If a list of dict/series is passed and the keys are all contained in
+            the DataFrame's index, the order of the columns in the resulting
+            DataFrame will be unchanged.
+            Iteratively appending rows to a cudf DataFrame can be more
+            computationally intensive than a single concatenate. A better
+            solution is to append those rows to a list and then concatenate
+            the list with the original DataFrame all at once.
+            `verify_integrity` parameter is not supported yet.
```
Comment (wence-) [inline python/cudf/cudf/core/dataframe.py:6712]: This is not a pandas compatibility note (in fact, the same note appears in the pandas docstring for this method), so can we restore to the previous state please?

File: python/cudf/cudf/core/dataframe.py [hunk 7]
```diff
@@ -7209,6 +7207,14 @@ def value_counts(
...
         6         0            1
         dtype: int64
+
+        .. pandas-compat::
+            **DataFrame.value_counts**
+
+            The returned Series will have a MultiIndex with one level per input
+            column. By default, rows that contain any NA values are omitted from
+            the result. By default, the resulting Series will be in descending
+            order so that the first element is the most frequently-occurring row.
```
Comment (wence-) [inline python/cudf/cudf/core/dataframe.py:7217]: This is not a pandas compatibility note (in fact, the same note appears in the pandas docstring for this method), so can we restore to the previous state please?

File: python/cudf/cudf/core/frame.py [hunk 8]
```diff
@@ -298,6 +293,12 @@ def empty(self):
         Series([], dtype: float64)
         >>> s.empty
         True
+
+        .. pandas-compat::
+            **DataFrame.empty, Series.empty, Frame.empty**
+
+            If DataFrame/Series contains only `null` values, it is still not
+            considered empty. See the example below.
```
Comment (wence-) [inline python/cudf/cudf/core/frame.py:301]:
```suggestion
            considered empty. See the example above.
```

File: python/cudf/cudf/core/groupby/groupby.py (hunk 1 above)
Comment (wence-) [reply python/cudf/cudf/core/groupby/groupby.py:1843]: I am -1 on this change since it violates OAOO for an only very marginal gain.

Reading the `Reducible` mixin, I think that `{cls}` will be formatted to the class name of the implementing class. So can we drop these additions, and modify the groupby change (see above).

File: python/cudf/cudf/core/groupby/groupby.py [hunk 9]
```diff
@@ -535,9 +535,9 @@ def _reduce(
         Series or DataFrame
             Computed {op} of values within each group.
 
-        Notes
-        -----
-        Difference from pandas:
+        .. pandas-compat::
+            **GroupBy.{op}**
```
Comment (wence-) [inline python/cudf/cudf/core/groupby/groupby.py:539]:
```suggestion
            **{cls}.{op}**
```

I think this will remove the need to replicate the `_reduce` method just to get the class name into the docstring correctly.

File: python/cudf/cudf/core/groupby/groupby.py [hunk 10]
```diff
@@ -1868,6 +1909,46 @@ def apply(self, func, *args):
...
+        """
+        if numeric_only:
+            raise NotImplementedError(
+                "numeric_only parameter is not implemented yet"
+            )
+        if min_count != 0:
+            raise NotImplementedError(
+                "min_count parameter is not implemented yet"
+            )
+        return self.agg(op)
```
Comment (wence-) [inline python/cudf/cudf/core/groupby/groupby.py:1950]: As for `DataFrameGroupBy`, with my proposed change to `GroupBy`, I think this can be dropped.

File: python/cudf/cudf/core/tools/datetimes.py [hunk 11]
```diff
@@ -809,6 +789,28 @@ def date_range(
                 '2026-04-23 08:00:00'],
                 dtype='datetime64[ns]')
 
+    .. pandas-compat::
+        **cudf.date_range**
+
+        Of the four parameters `start`, `end`, `periods`, and `freq`, exactly three
+        must be specified. If `freq` is omitted, the resulting DatetimeIndex will
+        have periods linearly spaced elements between start and end (closed on both
+        sides).
```
Comment (wence-) [inline python/cudf/cudf/core/tools/datetimes.py:798]: This part is not a pandas compatibility note, so please restore it to its previous location.

**Review by wence- — CHANGES_REQUESTED** [review]
Thanks! A few minor comments where I think the note is not actually a pandas compat issue.

File: python/cudf/cudf/core/groupby/groupby.py (hunk 1 above)
Comment (Touutae-lab) [reply python/cudf/cudf/core/groupby/groupby.py:1843]: Thank you for this, it took me so long to find how to change these I will experiment and commit change soon.

**Comment by vyasr** [comment]
@Touutae-lab could you retarget your changes to branch-23.04? We are no longer making changes for the 23.02 release. Thank you!

File: python/cudf/cudf/core/frame.py [hunk 12]
```diff
@@ -298,6 +293,12 @@ def empty(self):
         Series([], dtype: float64)
         >>> s.empty
         True
+
```
Comment (Touutae-lab) [inline python/cudf/cudf/core/frame.py:296]: this part still need improvement.

Comment (Touutae-lab) [reply python/cudf/cudf/core/frame.py:296]: Because many method in frame are used in Both DataFrame and Series the documentation will likely generate both for DataFrame and Series with the same class name that can lead to confusion.

Comment (Touutae-lab) [reply python/cudf/cudf/core/frame.py:296]: Because I'm lack understanding in `delegateMixin` `Operation` class so i resolve it like this. i will help change it later thank you.

Comment (Touutae-lab) [reply python/cudf/cudf/core/frame.py:296]: The alternative is add simple code such as `class.method.__doc__ = formatDocstring`

File: python/cudf/cudf/core/dataframe.py [hunk 13]
```diff
@@ -6704,7 +6707,7 @@ def append(
         1  1
         2  2
         3  3
-        4  4
+        4  4.
```
Comment (Touutae-lab) [inline python/cudf/cudf/core/dataframe.py:6710]:
```suggestion
        4  4
```

File: python/dask_cudf/dask_cudf/accessors.py [hunk 14]
```diff
@@ -277,6 +272,11 @@ def sort_values(
         1         [2.0, 8.0, 8.0]
         2              [1.0, 2.0]
         dtype: list
+
+        .. pandas-compat::
+            **ListMethods.sort_values**
+
+            * Not supporting: `inplace`, `kind`
```
Comment (Touutae-lab) [inline python/dask_cudf/dask_cudf/accessors.py:279]:
```suggestion
            The `inplace` and `kind` is currently unsupported.
```

File: python/cudf/cudf/core/series.py [hunk 15]
```diff
@@ -1996,6 +1991,12 @@ def sort_values(
         3    4
         1    5
         dtype: int64
+
+        .. pandas-compat::
+            **Series.sort_values**
+
+            * Support axis='index' only.
+            * Not supporting: inplace, kind
```
Comment (Touutae-lab) [inline python/cudf/cudf/core/series.py:1999]:
```suggestion
            The inplace and kind argument is currently unsupported
```

**Comment by Touutae-lab** [comment]
@vyasr of course ! i will keep update and merge. to keep updating with current `main` branch.

**Comment by vyasr** [comment]
@Touutae-lab did you intend to close both this PR and and #12691 (and delete the branch)?

**Comment by Touutae-lab** [comment]
No, I'm trying to fix commit.

File: python/cudf/cudf/core/groupby/groupby.py [hunk 16]
```diff
@@ -2151,6 +2151,28 @@ def fillna(
...
+                       v
+                k
+                1 0  2.0
+                  1  4.0
+                2 2  4.0
+                >>> gdf.groupby('k').fillna({'v': 4}) # cudf
+                     v
+                0  2.0
+                1  4.0
+                2  4.0
```
Comment (wence-) [inline python/cudf/cudf/core/groupby/groupby.py:2175]: I don't see this difference:
```
In [1]: import pandas as pd

In [2]: import cudf
df 
In [3]: df = pd.DataFrame({'k': [1, 1, 2], 'v': [2, None, 4]})

In [4]: gdf = cudf.from_pandas(df)

In [5]: df.groupby('k').fillna({'v': 4})
Out[5]: 
     v
0  2.0
1  4.0
2  4.0

In [6]: gdf.groupby('k').fillna({'v': 4})
Out[6]: 
     v
0  2.0
1  4.0
2  4.0
```
So:
```suggestion
```

**Review by wence- — DISMISSED** [review]
Thanks for keeping at this. I think there is just one note in addition to those that @vyasr commented on that I am confused by.

**Comment by Touutae-lab** [comment]
I will update it tmr, I have full time job now so I have to work on weekend

**Comment by Touutae-lab** [comment]
Updated And already build and has some manual check

**Comment by wence-** [comment]
Can you please fix the style-check errors, you can see them locally by doing `pre-commit run --all-files` in your checkout.

**Comment by Touutae-lab** [comment]
@wence- Ok, i will change wording or ETC to reduce line of document

File: python/cudf/cudf/tests/test_dataframe.py [hunk 17]
```diff
@@ -10068,82 +10068,3 @@ def test_dataframe_transpose_complex_types(data):
...
-            {
-                "a": [1, 2, 3],
-                "b": ["x", "y", "z"],
-            },
-            [10, 11],
-        ),
-        ([[10, 11], [12, 13]], ["a", "b", "c"]),
-    ],
-)
-def test_dataframe_init_length_error(data, index):
```
Comment (galipremsagar) [inline python/cudf/cudf/tests/test_dataframe.py:10137]: @Touutae-lab Curious as to why we had to remove all these pytests in `test_dataframe.py`? All I see is docstring fixes in this PR no?

**Review by galipremsagar — CHANGES_REQUESTED** [review]

**Review by vyasr — DISMISSED** [review]
@Touutae-lab can you undo the removal of the tests in tests/test_dataframe.py, merge in the latest 23.10, and retarget the PR at 23.10 (23.08 is now released)? After that I think this PR should be ready to go! I don't want to let it sit too much longer, you've done great work keeping it up to date and we would definitely benefit from getting it merged soon. Thanks for sticking with it!

File: python/cudf/cudf/tests/test_dataframe.py
[the remark this answers is not in the document]
Comment (vyasr) [reply python/cudf/cudf/tests/test_dataframe.py:1]: I've fixed this.

**Comment by vyasr** [comment]
@Touutae-lab I fixed a couple of things up for you here and I think this PR is good to go now. Sorry for the delay, and thank you for the contribution!

**Review by vyasr — APPROVED** [review]
